homeassistant/components/eq3btsmart/climate.py

```diff
@@ -0,0 +1,391 @@
+            manufacturer=MANUFACTURER,
+            model=DEVICE_MODEL,
+            identifiers={(DOMAIN, self._eq3_config.mac_address)},
+            sw_version=self._firmware_version,
+            connections={(CONNECTION_BLUETOOTH, self._eq3_config.mac_address)},
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+
+        asyncio.get_event_loop().create_task(self._async_scan_loop())
+        self._on_connection_changed(True)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when entity will be removed from hass."""
+
+        if self._cancel_timer:
+            self._cancel_timer()
+
+    async def _async_scan_loop(self, now: datetime | None = None) -> None:
+        """Scan for data."""
+
+        await self.async_scan()
+
+        if self._platform_state != EntityPlatformState.REMOVED:
+            delay = timedelta(seconds=self._eq3_config.scan_interval)
+            self._cancel_timer = async_call_later(
+                self.hass, delay, self._async_scan_loop
+            )
+
+    @callback
+    def _on_updated(self) -> None:
+        """Handle updated data from the thermostat."""
+
+        self._is_available = True
+
+        if self._thermostat.status is not None:
+            self._on_status_updated()
+
+        if self._thermostat.device_data is not None:
+            self._on_device_updated()
+
+        self.schedule_update_ha_state()
+
+    @callback
+    def _on_status_updated(self) -> None:
+        """Handle updated status from the thermostat."""
+
+        self._target_temperature = self._thermostat.status.target_temperature.value
+        self._attr_hvac_mode = EQ_TO_HA_HVAC[self._thermostat.status.operation_mode]
+
+    @callback
+    def _on_device_updated(self) -> None:
+        """Handle updated device data from the thermostat."""
+
+        self._firmware_version = str(self._thermostat.device_data.firmware_version)
+        self._attr_device_info["sw_version"] = self._firmware_version
+
+        device_registry = async_get(self.hass)
+        device = device_registry.async_get_device(
+            identifiers={(DOMAIN, self._eq3_config.mac_address)},
+        )
+
+        if device:
+            device_registry.async_update_device(
+                device.id,
+                sw_version=self._firmware_version,
+            )
+
+    @callback
+    def _on_connection_changed(self, is_connected: bool = True) -> None:
+        """Handle connection changed."""
+
+        self._is_available = is_connected
+
+        if is_connected and not self._was_connected:
+            _LOGGER.info("[%s] Connected", self._eq3_config.name)
+            self._was_connected = True
+            self.hass.add_job(self._update_device())
+
+        if not is_connected and self._was_connected:
+            _LOGGER.warning("[%s] Disconnected", self._eq3_config.name)
+            self._was_connected = False
+
+        self.schedule_update_ha_state()
```

Comment:
Please make sure you are using the async api in `@callback` and `async` functions.

https://github.com/home-assistant/core/blob/52d27230bce239017722d8ce9dd6f5386f63aba2/homeassistant/helpers/entity.py#L1207